horde_worker_regen/process_management/process_manager.py

```diff
@@ -2002,13 +2115,33 @@ async def _bridge_data_loop(self) -> None:
             except CancelledError:
                 self._shutting_down = True
 
+    @staticmethod
+    def _handle_exception(future: asyncio.Future) -> None:
+        """
+        Logs exceptions from asyncio tasks.
+
+        :param future: asyncio task to monitor
+        :return: None
+        """
+        ex = future.exception()
+        if ex is not None:
+            logger.error(f'exception thrown by a main loop task: {ex}')
+
     async def _main_loop(self) -> None:
         # Run both loops concurrently
+        process_control_loop = asyncio.create_task(self._process_control_loop(), name="process_control_loop")
+        api_call_loop = asyncio.create_task(self._api_call_loop(), name="api_call_loop")
+        job_submit_loop = asyncio.create_task(self._job_submit_loop(), name="job_submit_loop")
+        bridge_data_loop = asyncio.create_task(self._bridge_data_loop(), name="bridge_data_loop")
+        process_control_loop.add_done_callback(self._handle_exception)
```

Comment:
I do not have a huge problem with handling the exceptions this way, but I would point you to https://docs.python.org/3.11/library/asyncio-task.html#id7 where you can see you can have `gather` return the exception as the return value which could then easily be checked.

In terms of using tasks here, I remain unclear if there is a benefit, as all of these functions run in loops and are not execution order dependent.  Further, I do not foresee the need to cancel these functions from running (another possible benefit of tasks). I am also not entirely clear on the intention because my understanding is that the done callback is only happening if the task has completely returned, at which point (I think) the program is either ending or unrecoverable.

Can you give any insight on the choice here? I am not a python SME, so perhaps I am not understanding something fully.

Reply:
Alright, I did sense the last commit was in response to something not behaving as expected. If you have any questions about design decisions or unclear logic in my code, as always, feel free to ask.